modem: modem_cmux: Quality improvements

bjarki-andreasen opened the pull request:
This PR provides a handful of improvements.

Commit:
modem: modem_cmux: Make async connect/disconnect stateful

This commit adds a check to the async connect and disconnect
functions to validate the CMUX is not already connected
or disconnected respectively. This was already part of the
sync connect and disconnect functions, so the sync functions
now simply wrap the async functions to avoid duplicate code.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
modem: modem_cmux: Reset CMUX events on release

This commit resets the CMUX events to match the
initial value of disconnected.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
modem: modem_cmux: Use k_event_test instead of K_NO_WAIT

This commit replaces the k_event_wait calls using K_NO_WAIT
with k_event_test.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
modem: modem_cmux: Clear DLCI pipes on open

This commit resets the receive ring buffer for each DLCI
pipe when they are opened. They may have old data stored
from last time they where opened.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
modem: modem_cmux: Improve logging

This commit adds logging for CMUX frames and commands and
their data, for both transmit and receive. It also removes
the superseded LOG_DBG() lines like "Received frame".

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
modem: modem_cellular: Reset receive state on attach

This commit resets the state of the CMUX receive state
machine when the CMUX instance is attached to a pipe.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

ycsin commented on subsys/modem/modem_cmux.c:
```diff
@@ -452,7 +533,7 @@ static void modem_cmux_on_dlci_frame(struct modem_cmux *cmux)
 
 	default:
-		modem_cmux_log_unknown_frame(cmux);
+		LOG_WRN("Unknown DLCI frame type");
```
```suggestion
		LOG_WRN("Unknown %s frame type", "DLCI");
```

ycsin commented on subsys/modem/modem_cmux.c:
```diff
@@ -373,7 +449,7 @@ static void modem_cmux_on_control_frame(struct modem_cmux *cmux)
 
 	default:
-		modem_cmux_log_unknown_frame(cmux);
+		LOG_WRN("Unknown control frame type");
```
```suggestion
		LOG_WRN("Unknown %s frame type", "control");
```

ycsin reviewed (commented):
not really a thorough review, just spotted some potential bytes savings

Commit:
modem: cmux: Optimize log message to save ROM

This commit reuses the string "Unknown %s frame type" for two
log messages, as suggested by ycsin.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

bjarki-trackunit commented:
Added optimizations requested by @ycsin, and created a log entry specifically for the `MODEM_CMUX` module since it produces an immense amount of log output

ycsin reviewed (approved):
Ideally the commit to optimize the LOG messages should be squashed with the one that introduced them.

I would probably also omit the _"This commit ..."_ in all the commit messages, they seem somewhat redundant
